Author: Christoph Feck <christoph@maxiom.de>
Origin: http://websvn.kde.org/?view=revision&revision=1255297
Description: Fix libgadu pthread check.
 Using grep in LIBGADU_INCLUDE_DIR can fail, if libgadu.h is
 located in /usr/include, but LIBGADU_INCLUDE_DIR only contains
 the additional required include paths, such as those added by gnutls,
 as found with latest openSUSE libgadu.pc

--- /dev/null
+++ b/kopete/cmake/modules/gadu_pthread_check.c
@@ -0,0 +1,11 @@
+#include <libgadu.h>
+
+int main(int argc, char *argv[])
+{
+#ifdef GG_CONFIG_HAVE_PTHREAD
+    return 1;
+#else
+    return 0;
+#endif
+}
+
--- a/kopete/cmake/modules/FindLibgadu.cmake
+++ b/kopete/cmake/modules/FindLibgadu.cmake
@@ -65,23 +65,35 @@ if (NOT LIBGADU_INCLUDE_DIR OR NOT LIBGA
     mark_as_advanced (LIBGADU_INCLUDE_DIR LIBGADU_LIBRARIES)
 
 endif (NOT LIBGADU_INCLUDE_DIR OR NOT LIBGADU_LIBRARIES)
-    
+
 if (LIBGADU_INCLUDE_DIR AND LIBGADU_LIBRARIES)
 
     if (NOT WIN32)
 
-        execute_process ( COMMAND grep "#define GG_CONFIG_HAVE_PTHREAD" ${LIBGADU_INCLUDE_DIR}/libgadu.h RESULT_VARIABLE LIBGADU_PTHREADS )
-    
-        if (${LIBGADU_PTHREADS} GREATER 0)
+        try_run (run_result compile_result ${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/cmake/modules/gadu_pthread_check.c
+            CMAKE_FLAGS -DINCLUDE_DIRECTORIES:STRING=${LIBGADU_INCLUDE_DIR}
+            COMPILE_DEFINITIONS "${LIBGADU_DEFINITIONS}"
+        )
 
-            MESSAGE(STATUS "libgadu must be compiled with pthreads support")
-            set (LIBGADU_FOUND FALSE)
+        if (compile_result)
+
+            if (run_result EQUAL 1)
+
+                set (LIBGADU_FOUND TRUE)
 
-        else (${LIBGADU_PTHREADS} GREATER 0)
+            else (run_result EQUAL 1)
 
-            set (LIBGADU_FOUND TRUE)
+                MESSAGE(STATUS "libgadu must be compiled with pthreads support")
+                set (LIBGADU_FOUND FALSE)
+
+            endif (run_result EQUAL 1)
+
+        else (compile_result)
+
+            MESSAGE(STATUS "unable to compile libgadu pthread check")
+            set (LIBGADU_FOUND FALSE)
 
-        endif (${LIBGADU_PTHREADS} GREATER 0)
+        endif (compile_result)
 
     else (NOT WIN32)
 
@@ -89,4 +101,4 @@ if (LIBGADU_INCLUDE_DIR AND LIBGADU_LIBR
 
     endif (NOT WIN32)
 
-endif (LIBGADU_INCLUDE_DIR AND LIBGADU_LIBRARIES)
\ No newline at end of file
+endif (LIBGADU_INCLUDE_DIR AND LIBGADU_LIBRARIES)
